Add `prototype_build_details_comment` action

AliSoftware opened the pull request:
## What does it do?

Many of our client apps are configured to generate a Prototype Build (aka Installable Build), then make a comment on the PR with a QR code and some additional info to download the build directly from App Center.

Since the HTML content of those comments are a bit annoying and verbose to hardcode directly in the Fastfile of each of our apps, while they all look the same, I've decided to DRY the generation of this content in a dedicated action that generates a nice HTML table (with the QR code + some metadata) for us, ensuring we use the same nice format in all our client apps.

## Checklist before requesting a review

- [x] Run `bundle exec rubocop` to test for code style violations and recommendations
- [x] Add Unit Tests (aka `specs/*_spec.rb`) if applicable
- [x] Run `bundle exec rspec` to run the whole test suite and ensure all your tests pass
- [x] Make sure you added an entry in [the `CHANGELOG.md` file](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/CHANGELOG.md#trunk) to describe your changes under the approprioate existing `###` subsection of the existing `## Trunk` section.

AliSoftware committed:
Add `prototype_build_details_comment` action

AliSoftware committed:
Add CHANGELOG entry

AliSoftware committed:
Don't wrap arbitrary metadata in `<tt>` tags

AliSoftware committed:
Add ability to hide the table behind `<details>`

AliSoftware committed:
Add ability to provide a direct download link

Useful for APK files for example.

PS: be mindful that providing a public link accessible without any access control (e.g. from a cloudfront public API or similar) means users outside Automattic might be able to download the APK even if they are not invited to App Center. Up to you to decide if that is ok or not when you choose to use this parameter.

AliSoftware committed:
Fix `<details>` block rendering/syntax

AliSoftware committed:
Shake up some tests to vary tested cases more

AliSoftware committed:
Refine table key names

AliSoftware committed:
Shorten Intro text

AliSoftware commented on spec/prototype_build_details_comment_action_spec.rb:
```diff
@@ -0,0 +1,173 @@
+        footnote: '<em>Note: Google Sign-In in not available in those builds</em>'
+      )
+
+      expect(comment).to eq <<~EXPECTED_COMMENT
+        <p>📲 You can test the changes from this Pull Request in <strong>BestApp</strong> by scanning the QR code below to install the corresponding build via App Center.</p>
+        <table>
+        <tr>
+          <td rowspan='6'><img src='https://chart.googleapis.com/chart?chs=500x500&cht=qr&chl=https%3A%2F%2Finstall.appcenter.ms%2Forgs%2FBestOrg%2Fapps%2FBestApp%2Freleases%2F8888&choe=UTF-8' width='250' height='250' /></td>
+          <td width='150px'><b>App Name</b></td><td><tt>BestApp</tt></td>
+        </tr>
+        <tr><td><b>Version:Short</b></td><td>28.2</td></tr>
+        <tr><td><b>Version:Long</b></td><td>28.2.0.108</td></tr>
+        <tr><td><b>Flavor</b></td><td>Celray</td></tr>
+        <tr><td><b>App Center Build</b></td><td><a href='https://install.appcenter.ms/orgs/BestOrg/apps/BestApp/releases/8888'>Build \#8888</a></td></tr>
+        <tr><td><b>Commit</b></td><td><tt>a1b2c3f</tt></td></tr>
+        </table>
+        <em>Note: Google Sign-In in not available in those builds</em>
+      EXPECTED_COMMENT
```
FYI, this is how that example looks like once rendered:<hr/>

<p>📲 You can test the changes from this Pull Request in <strong>BestApp</strong> by scanning the QR code below to install the corresponding build via App Center.</p>
<table>
<tr>
  <td rowspan='6'><img src='https://chart.googleapis.com/chart?chs=500x500&cht=qr&chl=https%3A%2F%2Finstall.appcenter.ms%2Forgs%2FBestOrg%2Fapps%2FBestApp%2Freleases%2F8888&choe=UTF-8' width='250' height='250' /></td>
  <td width='150px'><b>App Name</b></td><td><tt>BestApp</tt></td>
</tr>
<tr><td><b>Version:Short</b></td><td>28.2</td></tr>
<tr><td><b>Version:Long</b></td><td>28.2.0.108</td></tr>
<tr><td><b>Flavor</b></td><td>Celray</td></tr>
<tr><td><b>App Center Build</b></td><td><a href='https://install.appcenter.ms/orgs/BestOrg/apps/BestApp/releases/8888'>Build #8888</a></td></tr>
<tr><td><b>Commit</b></td><td><tt>a1b2c3f</tt></td></tr>
</table>
<em>Note: Google Sign-In in not available in those builds</em>

AliSoftware commented on spec/prototype_build_details_comment_action_spec.rb:
```diff
@@ -0,0 +1,173 @@
+        metadata: metadata
+      )
+
+      expect(comment).to eq <<~EXPECTED_COMMENT
+        <p>📲 You can test the changes from this Pull Request in <strong>BestApp</strong> by scanning the QR code below to install the corresponding build via App Center.</p>
+        <table>
+        <tr>
+          <td rowspan='6'><img src='https://chart.googleapis.com/chart?chs=500x500&cht=qr&chl=https%3A%2F%2Finstall.appcenter.ms%2Forgs%2FBestOrg%2Fapps%2FBestApp%2Freleases%2F8888&choe=UTF-8' width='250' height='250' /></td>
+          <td width='150px'><b>App Name</b></td><td><tt>BestApp</tt></td>
+        </tr>
+        <tr><td><b>Version:Short</b></td><td>28.2</td></tr>
+        <tr><td><b>Version:Long</b></td><td>28.2.0.108</td></tr>
+        <tr><td><b>Direct Link</b></td><td><a href='https://bestfront.cloudfront.net/feed42/bestapp-pr1357-a1b2c3f.apk'><tt>bestapp-pr1357-a1b2c3f.apk</tt></a></td></tr>
+        <tr><td><b>App Center Build</b></td><td><a href='https://install.appcenter.ms/orgs/BestOrg/apps/BestApp/releases/8888'>Build \#8888</a></td></tr>
+        <tr><td><b>Commit</b></td><td><tt>a1b2c3f</tt></td></tr>
+        </table>
+        <em>Automatticians: You can use our internal self-serve MC tool to give yourself access to App Center if needed.</em>
+      EXPECTED_COMMENT
```
FYI, this is how that example looks like once rendered:<hr/>

<p>📲 You can test the changes from this Pull Request in <strong>BestApp</strong> by scanning the QR code below to install the corresponding build via App Center.</p>
<table>
<tr>
  <td rowspan='6'><img src='https://chart.googleapis.com/chart?chs=500x500&cht=qr&chl=https%3A%2F%2Finstall.appcenter.ms%2Forgs%2FBestOrg%2Fapps%2FBestApp%2Freleases%2F8888&choe=UTF-8' width='250' height='250' /></td>
  <td width='150px'><b>App Name</b></td><td><tt>BestApp</tt></td>
</tr>
<tr><td><b>Version:Short</b></td><td>28.2</td></tr>
<tr><td><b>Version:Long</b></td><td>28.2.0.108</td></tr>
<tr><td><b>Direct Link</b></td><td><a href='https://bestfront.cloudfront.net/feed42/bestapp-pr1357-a1b2c3f.apk'><tt>bestapp-pr1357-a1b2c3f.apk</tt></a></td></tr>
<tr><td><b>App Center Build</b></td><td><a href='https://install.appcenter.ms/orgs/BestOrg/apps/BestApp/releases/8888'>Build #8888</a></td></tr>
<tr><td><b>Commit</b></td><td><tt>a1b2c3f</tt></td></tr>
</table>
<em>Automatticians: You can use our internal self-serve MC tool to give yourself access to App Center if needed.</em>

AliSoftware commented on spec/prototype_build_details_comment_action_spec.rb:
```diff
@@ -0,0 +1,173 @@
+        footnote: '<em>Note: Google Sign-In in not available in those builds</em>'
+      )
+
+      expect(comment).to eq <<~EXPECTED_COMMENT
+        <details><summary>📲 You can test the changes from this Pull Request in <strong>BestApp</strong> by scanning the QR code below to install the corresponding build via App Center.</summary>
+        <table>
+        <tr>
+          <td rowspan='7'><img src='https://chart.googleapis.com/chart?chs=500x500&cht=qr&chl=https%3A%2F%2Finstall.appcenter.ms%2Forgs%2FBestOrg%2Fapps%2FBestApp%2Freleases%2F1234&choe=UTF-8' width='250' height='250' /></td>
+          <td width='150px'><b>App Name</b></td><td><tt>BestApp</tt></td>
+        </tr>
+        <tr><td><b>Version:Short</b></td><td>28.2</td></tr>
+        <tr><td><b>Version:Long</b></td><td>28.2.0.108</td></tr>
+        <tr><td><b>Flavor</b></td><td>Celray</td></tr>
+        <tr><td><b>Configuration</b></td><td>Debug</td></tr>
+        <tr><td><b>App Center Build</b></td><td><a href='https://install.appcenter.ms/orgs/BestOrg/apps/BestApp/releases/1234'>Build \#1234</a></td></tr>
+        <tr><td><b>Commit</b></td><td><tt>a1b2c3f</tt></td></tr>
+        </table>
+        <em>Note: Google Sign-In in not available in those builds</em>
+        </details>
+      EXPECTED_COMMENT
```
FYI, this is how that example looks like once rendered:<hr/>

<details><summary>📲 You can test the changes from this Pull Request in <strong>BestApp</strong> by scanning the QR code below to install the corresponding build via App Center.</summary>
<table>
<tr>
  <td rowspan='7'><img src='https://chart.googleapis.com/chart?chs=500x500&cht=qr&chl=https%3A%2F%2Finstall.appcenter.ms%2Forgs%2FBestOrg%2Fapps%2FBestApp%2Freleases%2F1234&choe=UTF-8' width='250' height='250' /></td>
  <td width='150px'><b>App Name</b></td><td><tt>BestApp</tt></td>
</tr>
<tr><td><b>Version:Short</b></td><td>28.2</td></tr>
<tr><td><b>Version:Long</b></td><td>28.2.0.108</td></tr>
<tr><td><b>Flavor</b></td><td>Celray</td></tr>
<tr><td><b>Configuration</b></td><td>Debug</td></tr>
<tr><td><b>App Center Build</b></td><td><a href='https://install.appcenter.ms/orgs/BestOrg/apps/BestApp/releases/1234'>Build #1234</a></td></tr>
<tr><td><b>Commit</b></td><td><tt>a1b2c3f</tt></td></tr>
</table>
<em>Note: Google Sign-In in not available in those builds</em>
</details>

AliSoftware committed:
Provide additional doc for appcenter_release_id

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/prototype_build_details_comment_action.rb:
```diff
@@ -0,0 +1,130 @@
+        install_url = "https://install.appcenter.ms/orgs/#{appcenter_org_name}/apps/#{appcenter_app_name}/releases/#{appcenter_release_id}"
+        qr_code_url = "https://chart.googleapis.com/chart?chs=500x500&cht=qr&chl=#{CGI.escape(install_url)}&choe=UTF-8"
+        metadata_rows = metadata.compact.map do |key, value|
+          "<tr><td><b>#{key}</b></td><td>#{value}</td></tr>"
+        end
+
+        intro = "📲 You can test the changes from this Pull Request in <strong>#{appcenter_app_name}</strong> by scanning the QR code below to install the corresponding build via App Center."
```
I noticed the usage of `<b>` in the table rows and `<strong>` in this intro text.

That sent me into a mini-rabbit-hole to understand the difference between the two. I found the [`<strong>` documentation from Mozilla](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/strong#b_vs._strong) useful:

> The `<strong>` element is for content that is of greater importance, while the `<b>` element is used to draw attention to text without indicating that it's more important.
>
> It may help to realize that both are valid and semantic elements in HTML and that it's a coincidence that they both have the same default styling (boldface) in most browsers (although some older browsers actually underline `<strong>`). 

With that in mind, I'd suggest making this `<b>`. At the same time, I don't think we need to sweat the HTML semantics for the content of a comment going in a GitHub PR...

AliSoftware replied:
Addressed in 4204a4532477907f124a32b01f39cccddcfdece9

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/prototype_build_details_comment_action.rb:
```diff
@@ -0,0 +1,130 @@
+          <table>
+          <tr>
+            <td rowspan='#{metadata.count + 3}'><img src='#{qr_code_url}' width='250' height='250' /></td>
+            <td width='150px'><b>App Name</b></td><td><tt>#{appcenter_app_name}</tt></td>
```
[`<tt>` has been depracted](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/tt).

What do you think of using `<pre>`?

AliSoftware replied:
Seems like `<pre>` is a `display:block` element (like a `div`) aimed to stylize an entire block of potentially-multiline text (e.g. ASCII art), while here we might prefer to keep an `display:inline` element (like a `span`) that can be applied to only a portion of the paragraph. So I think `<code>` would be a more appropriate replacement for that `<tt>`

AliSoftware replied:
Addressed in 911d71bb02533ba7cc618e9f7f1490789d4b6f30

mokagio replied:
Ah, nice! HTML is cool like that.

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/prototype_build_details_comment_action.rb:
```diff
@@ -0,0 +1,130 @@
+      #####################################################
+
+      def self.description
+        'Generates the nicely-formatted string providing all the details of a prototype build, ready to be used as a PR comment (e.g. via `comment_on_pr`).'
```
Nitpick: "Generates _the_ nicely-formatted" string..." makes me think we've just been talking about it, but a reader might not have context on it when checking the documentation.

Also, is the string that is nicely formatted, or is it the HTML content it contain?

```suggestion
        'Generates a string providing all the details of a prototype build, nicely-formatted and ready to be used as a PR comment (e.g. via `comment_on_pr`).'
```

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/prototype_build_details_comment_action.rb:
```diff
@@ -0,0 +1,130 @@
+          ),
+          FastlaneCore::ConfigItem.new(
+            key: :commit,
+            env_name: 'BUILDKITE_COMMIT',
```
Worth pointing out that this adds an implicit dependency on Buildkite as the CI provider.

It's a pretty safe assumption to make, given this is an internal tool only.

AliSoftware replied:
Tbh I initially thought about writing this as `env_names: ['BUILDKITE_COMMIT', …]` and add support for the env vars used by more CI providers. But I figured that I had other fishes to fry to chase down the env vars used by different CI providers while we wouldn't need it internally.

I'm still gonna add a comment to make that clearer though.

AliSoftware replied:
Updated with comments in 854098f3e5995f3e45eaa5b893ed8c50e906a2aa

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/prototype_build_details_comment_action.rb:
```diff
@@ -0,0 +1,130 @@
+          ),
+          FastlaneCore::ConfigItem.new(
+            key: :commit,
+            env_name: 'BUILDKITE_COMMIT',
```
Worth adding a link to the docs, just in case?

```suggestion
            env_name: 'BUILDKITE_COMMIT', # https://buildkite.com/docs/pipelines/environment-variables#BUILDKITE_COMMIT
```

AliSoftware replied:
Added as part of 854098f3e5995f3e45eaa5b893ed8c50e906a2aa

mokagio commented on spec/prototype_build_details_comment_action_spec.rb:
```diff
@@ -0,0 +1,173 @@
+  end
+
+  describe 'expected info is included' do
+    it 'generates the proper AppCenter link and QR code given an org, app name and release ID' do
```
Nitpick
```suggestion
    it 'generates the proper App Center link and QR code given an org, app name and release ID' do
```

mokagio reviewed (approved):
Thank you for this. It's a great DRY improvement. Leverage++

None of my comments are blockers but I think it would be nice to integrate this in a real repo as a test before merging. That work wouldn't be wasted, as could be updated into a PR to integrate the stable version of these changes once released.

AliSoftware committed:
s/<strong>/<b>/

As suggested in https://github.com/wordpress-mobile/release-toolkit/pull/449\#discussion_r1093903902

AliSoftware commented:
> I think it would be nice to integrate this in a real repo as a test before merging. That work wouldn't be wasted, as could be updated into a PR to integrate the stable version of these changes once released.

Yep, that has actually already been started… but on one of the Tumblr repo in our GH:E instance https://github.tumblr.net/TumblrMobile/orangina/pull/23867 🙃 
I'm planning to make that Tumblr PR point to this branch and start using both the actions I've added as part of #442 + the one added here to refine that PR; once I've proved this worked well for that Tumblr repo, I'm planning on landing this then replicating the improvement to all our client app repos (then P2 about it) some time next sprint.

AliSoftware committed:
s/<tt>/<code>/

As discussed in https://github.com/wordpress-mobile/release-toolkit/pull/449\#discussion_r1095536151

AliSoftware committed:
Improve wording on action description

Co-authored-by: Gio Lodi <gio.lodi@automattic.com>

AliSoftware committed:
Add comment on the Buildkite-specificity of env var used by the `:commit` `ConfigItem`

As discussed in https://github.com/wordpress-mobile/release-toolkit/pull/449\#discussion_r1093910965

AliSoftware committed:
Typo on spec name

Co-authored-by: Gio Lodi <gio.lodi@automattic.com>

AliSoftware commented:
Testing by starting to use it in a Draft PR in WPiOS: https://github.com/wordpress-mobile/WordPress-iOS/pull/20071

mokagio commented:
@AliSoftware 

> Yep, that has actually already been started… but on one of the Tumblr repo in our GH:E instance [github.tumblr.net/TumblrMobile/orangina/pull/23867](https://github.tumblr.net/TumblrMobile/orangina/pull/23867) 🙃

Of course 🤦‍♂️ I didn't put the two things together, should have expected it. Still, thank you for opening a test PR in the WordPress iOS repo as well. _Hash-tag working in public, I guess? 🤷‍♂️ 😅_

AliSoftware committed:
Remove <code> around app name and commit

Especially, leaving the commit bare without <code> tag will allow GitHub to automatically transform it into a link to the commit itself

AliSoftware committed:
Refactor the action to auto-extract most of the info

From values provided in the `lane_context` by the `appcenter_upload` action – if it was used to distribute the Prototype build

AliSoftware commented:
Currently in testing in the following repos to tweak the latest adjustments needed:

 - [Tumblr-Android](https://github.tumblr.net/TumblrMobile/android2/pull/19647): APK uploaded to S3, not App Center
 - [Tumblr-iOS](https://github.tumblr.net/TumblrMobile/orangina/pull/23867): Build uploaded to App Center from a Mac agent… **but** PR comment created via separate CI job run on a different agent (one which has access to Tumblr's private network and GH:E instance), meaning `lane_context` can't be magically used by the action
 - [WordPress-Android](https://github.com/wordpress-mobile/WordPress-Android/pull/17926) (APK uploaded to S3, download URL exposed via Cloudfront)
 - [WordPress-iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/20071): Build uploaded to App Center, and PR comment created in the same lane, allowing us to have the action use `lane_context` 🎉 

> **Note**: Tests are currently out of sync with the recent refactoring of the API, so will fail. This is expected, as I wanted to focus on adjusting the final API and trying it out in the various client repos above first, before spending the time adjusting the specs only once the API would be confirmed to be fit for all client app's cases above.

TODO:
 - [x] Finish testing in the above 4 repos to validate the latest API of this action is a good fit and renders as expected in all the cases
 - [x] Update the unit tests (specs) to match the new confirmed API
 - [x] Fix the rubocop violations (then ensure the test still pass after adjusting the code to fix them)
 - [x] Test one last time in the repos above with the latest changes in the action
 - [x] Re-request review, get it approved, release a new version of the toolkit
 - [ ] Update the Draft PRs above to point to the released version of the toolkit
 - [ ] P2 it, and profit

AliSoftware committed:
Fix existing tests

AliSoftware commented:
Whooops, @mokagio I just saw that I re-requested your review on this PR by accident, while it's not yet ready to re-review (API have been confirmed to be a good fit for client apps, but tests haven't yet been updated with the new API yet nor expanded with all additional use cases… still working on that). And I'm not sure I can cancel a request for re-review on GitHub 🤔

AliSoftware committed:
Improving tests, pass 1

There are still more tests that I want to add, and I need to proof-read the ones that I wrote in this commit to ensure I didn't miss cases of adjusting the test content after copy/pasting code from various tests around.

mokagio commented:
No worries @AliSoftware .

AliSoftware committed:
Improving tests, pass 2

AliSoftware committed:
Split code in helpers to reduce complexity

AliSoftware committed:
Refactoring to split code in more separate methods

AliSoftware commented:
@mokagio The PR is now ready for final review again 🙂 . Lots of small refactorings and lots of additional unit tests since your last review 😉

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/prototype_build_details_comment_action.rb:
```diff
@@ -0,0 +1,259 @@
+          FastlaneCore::ConfigItem.new(
+            key: :app_center_org_name,
+            env_name: 'APPCENTER_OWNER_NAME', # Intentionally the same as the one used by the `appcenter_upload` action
+            description: 'The name of the organization in App Center (if you used `appcenter_upload` to distribute your Prototype build)',
```
```suggestion
            description: 'The name of the organization in App Center #{app_center_auto}',
```

AliSoftware replied:
Nope, in this case it was intentional not to use `{app_center_auto}` (aka "(will be automatically extracted from `lane_context if you used `appcenter_upload` to distribute your Prototype build)"), because that's not the case for this property.

In fact, the `app_center_org_name` is the only value that you have to provide explicitly if you want the action to use App Center.
 - Not only is this value not exposed in the `lane_context` filled by `appcenter_upload`, unlike all the other properties (the App Center API response includes the app name, but not the app org… who knows why), so this is NOT extracted from `lane_context` here
 - But also, I figured this was a great thing to have a way to tell the action to NOT assume that because `appcenter_upload` was used that means we want to use its `lane_context` implicitly (without any way of disabling that) for the comment generation.
    - For example, imagine a case where a fastlane lane (1) creates a beta build and uploads it App Center via `appcenter_upload` (2) and then creates a Prototype Build, which they do NOT upload to App Center but instead to S3. We don't want the `prototype_build_details_comment` lane to pick up the `lane_context` from the `appcenter_upload` call from (1) while we didn't use App Center for (2).
    - So in the end I decided to consider this fact of "`app_center_org_name` not being exposed in the `lane_context`" as a positive rather than a negative, as it allows you too use it as a signaling flag of it you want the action to consider you used App Center or not (instead not having a way to disable the `lane_context` inference) 🤷

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/prototype_build_details_comment_action.rb:
```diff
@@ -0,0 +1,259 @@
+          FastlaneCore::ConfigItem.new(
+            key: :app_icon,
+            env_name: 'FL_PROTOTYPE_BUILD_DETAILS_COMMENT_APP_ICON',
+            description: "An `:emojicode:` or URL to use for the icon of the app in the message. #{app_center_auto}",
```
TIL, [Emojicode is a programming language](https://www.emojicode.org/). A common name for the `:emoji:` (e.g. `: grinning_face:` =  😁) in the context of programming is "emoji short codes", while the official spec uses "[CLDR Short Name](https://unicode.org/emoji/format.html#col-name)".

However, neither applies here, truly, given the emojis are fetched from Buildkite 🤔. 

What do you think of this suggestion?

```suggestion
            description: "The name of an emoji from the https://github.com/buildkite/emojis list or the full image URL to use for the icon of the app in the message. #{app_center_auto}",
```

AliSoftware committed:
Improve description for `app_icon` parameter

Co-authored-by: Gio Lodi <gio.lodi@automattic.com>
